### PR TITLE
TextControl: remove margin overrides and add new opt-in prop (pt 2/2)

### PR DESCRIPTION
**Pull request by brookewp**
Part 2/2

## What?

Added new opt-in prop `__nextHasNoMarginBottom` for usages of `TextControl` in the Gutenberg codebase and removed margin overrides. 

This is part of https://github.com/WordPress/gutenberg/pull/47157 but was separated due to small visual changes to each of the blocks/components in this PR. 

## Why?

These changes will result in less CSS used in favour of components. 

Part of this project: https://github.com/WordPress/gutenberg/issues/38730
The tl;dr is `BaseControl` has a `margin-bottom` which makes it difficult to reuse and results in inconsistent use. 

## How?

By removing margin overrides in the CSS and adding the prop `__nextHasNoMarginBottom`. As well as adding `VStack` and `HStack` (and sometimes replacing `Flex` with those) to remove additional CSS.

## Additional Notes 

`packages/block-library/src/post-comment/edit.js` has been deprecated. I looked at it in the editor, and it looks a bit broken as it is: 

Screenshot 2023-01-11 at 12 16 16 PM

Adding the prop won't affect it by much, it shifts slightly down a pixel (similar to the legacy widget in t[he SelectControl PR](https://github.com/WordPress/gutenberg/pull/46448/files), which we updated without adding additional overrides), so I added it here to be consistent. 

## Testing Instructions

**Rather than verifying if the changes look the same as trunk, we'll need to make sure that these minor visual changes are okay or if we need to add CSS overrides to match perfectly to trunk.** 

#### Block Editor testing steps:

- <h3 id="HierarchicalTermSelector">HierarchicalTermSelector</h2>

	1. Edit post
	2. Look for 'Categories' in block inspector
	3. Change line 40 to a lower number if you don't have more than 8 categories on your test site: https://github.com/WordPress/gutenberg/blob/f7052d3e1e3e69f3b09b412c071c6c1eac69836f/packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js#L40
	4. Ensure the space below the text input for 'Search categories' looks the same as before 
	5. Click on 'Add New Category' 
	6. Ensure space below the text input for 'New category name' looks the same as before 
	<br />Screenshot 2023-01-10 at 7 21 11 PM

- ### NonDefaultControls (DateFormatPicker)

	1. Add a Post Date block
	2. Toggle off 'default format' in the block inspector
	3. Choose 'Custom' format from the dropdown
	5. Verify spacing 

	| <br /> Before  | <br />After |
	| --- | --- |
	| Screenshot 2023-01-13 at 10 54 57 PM | Screenshot 2023-01-13 at 10 55 02 PM |

- ### ImageURLInputUI

	1. Add an Image block and upload/add an image 
	2. Click on the Image block toolbar to add a URL
	3. Click on the downward-facing arrow to view additional options
	4. Verify spacing 

	| <br /> Before  | <br />After |
	| --- | --- |
	| Screenshot 2023-01-13 at 11 03 57 PM | Screenshot 2023-01-13 at 11 04 07 PM |

- ### SingleTrackEditor

	1. Add a Video block and upload/add video
	2. Click on 'Text tracks' in the toolbar
	3. Add or upload a .vtt text file (I used [a sample I found here](https://gist.github.com/samdutton/ca37f3adaf4e23679957b8083e061177))
	4. Verify spacing 
	
	| <br /> Before  | <br />After |
	| --- | --- |
	| Screenshot 2023-01-13 at 11 26 12 PM | Screenshot 2023-01-13 at 11 26 18 PM |

- ### SocialLinkEdit

	1. Add a Social Icons block to the editor
	2. Add any icon
	3. Look for 'Link label' in block inspector and see that there is a bit less space below, but it's more even
	4. Look for 'Link Rel' under 'Advanced'
	6. Ensure space below 'Link Rel' looks the same as trunk

	| <br /> Before  | <br />After |
	| --- | --- |
	| Screenshot 2023-01-11 at 10 37 06 PM | Screenshot 2023-01-11 at 10 37 11 PM |

<hr />

#### Site Editor testing steps:

- ### CreateTemplatePartModal

	1. Open site editor
	2. Click on template parts
	3. Click plus icon to create a new template part
	4. Verify spacing 

	| <br /> Before  | <br />After |
	| --- | --- |
	| Screenshot 2023-01-13 at 11 06 48 PM | Screenshot 2023-01-13 at 11 08 10 PM |

- ### TitleModal

	1. Edit a template in the site editor
	2. Add a Template Part block 
	3. Choose start blank
	4. Verify spacing 

	| <br /> Before  | <br />After |
	| --- | --- |
	| Screenshot 2023-01-13 at 11 10 46 PM | Screenshot 2023-01-13 at 11 09 40 PM |

<hr />

#### Storybook testing steps: 

- ### Placeholder

	1. `npm run storybook:dev`
	2. Open Placeholder component 
	3. Verify spacing 

	| <br /> Before  | <br />After |
	| --- | --- |
	| Screenshot 2023-01-13 at 11 13 51 PM | Screenshot 2023-01-13 at 11 14 12 PM |

<br />

### CODE REVIEW

File: packages/block-library/src/template-part/edit/title-modal.js [hunk 1]
```diff
@@ -33,16 +33,14 @@ export default function TitleModal( { areaLabel, onClose, onSubmit } ) {
...
-					label={ __( 'Name' ) }
-					value={ title }
-					onChange={ setTitle }
-				/>
-				<Flex
-					className="wp-block-template-part__placeholder-create-new__title-form-actions"
-					justify="flex-end"
-				>
-					<FlexItem>
+				<VStack spacing="5">
```
Comment (brookewp) [inline packages/block-library/src/template-part/edit/title-modal.js:36]: Added VStack with spacing to be consistent with other modals, as the spacing between the input and button wasn't the same as other modals in the site editor.

File: packages/edit-site/src/components/create-template-part-modal/index.js [hunk 2]
```diff
@@ -54,76 +55,84 @@ export default function CreateTemplatePartModal( { closeModal, onCreate } ) {
...
-					onChange={ setTitle }
-					required
-				/>
-				<BaseControl
-					label={ __( 'Area' ) }
-					id={ `edit-site-create-template-part-modal__area-selection-${ instanceId }` }
-					className="edit-site-create-template-part-modal__area-base-control"
-				>
-					<RadioGroup
+				<VStack spacing="4">
```
Comment (brookewp) [inline packages/edit-site/src/components/create-template-part-modal/index.js:58]: I personally think that the new spacing looks more centered, but I could be wrong. I can add CSS to make it like trunk if others prefer.

File: packages/edit-site/src/components/create-template-part-modal/style.scss [hunk 3]
```diff
@@ -8,16 +8,6 @@
...
 }
 
-
-.edit-site-create-template-part-modal__modal-actions {
-	padding-top: $grid-unit-15;
-}
-
-.edit-site-create-template-part-modal__area-base-control .components-base-control__label {
-	margin: $grid-unit-20 0 $grid-unit-10;
-	cursor: auto;
```
Comment (brookewp) [inline packages/edit-site/src/components/create-template-part-modal/style.scss:18]: Is there a reason why we'd use a text cursor here?

File: packages/block-editor/src/components/date-format-picker/style.scss [hunk 4]
```diff
@@ -1,18 +1,16 @@
 .block-editor-date-format-picker {
```
Comment (brookewp) [inline packages/block-editor/src/components/date-format-picker/style.scss]: We can add CSS to make it the same as trunk, but there is a 20px space so we'd either have to use 16px, 24px, or possibly a grid unit subtracted by another in order to get 20px if we're okay to not use a single grid unit.

File: packages/block-library/src/video/editor.scss [hunk 5]
```diff
@@ -71,36 +71,19 @@
...
-
 .block-library-video-tracks-editor__single-track-editor-kind-select {
 	max-width: 240px;
 }
 
 .block-library-video-tracks-editor__single-track-editor-buttons-container {
-	display: flex;
-	place-content: space-between;
-	margin-top: $grid-unit-40;
+	margin-top: $grid-unit-20;
```
Comment (brookewp) [inline packages/block-library/src/video/editor.scss]: Curious about when we should use a Spacer component instead of a margin like in this instance, but it seems like it could be debatable. :)

File: packages/components/src/placeholder/stories/index.tsx [hunk 6]
```diff
@@ -44,6 +44,7 @@ const Template: ComponentStory< typeof Placeholder > = ( args ) => {
 		<Placeholder { ...args }>
 			<div>
 				<TextControl
+					__nextHasNoMarginBottom
```
Comment (brookewp) [inline packages/components/src/placeholder/stories/index.tsx:47]: Adding to stories, as we did with Disabled in [a related PR](https://github.com/WordPress/gutenberg/pull/46559/files). The spacing is also similar to what we did with the legacy widget in [the SelectControl PR](https://github.com/WordPress/gutenberg/pull/46448/files),

File: packages/block-editor/src/components/date-format-picker/style.scss [hunk 7]
```diff
@@ -1,18 +1,16 @@
...
 .block-editor-date-format-picker__default-format-toggle-control__hint {
 	color: $gray-700;
 	display: block;
 }
 
-.block-editor-date-format-picker__custom-format-select-control {
-	&.components-base-control {
-		margin-bottom: 0;
-	}
-}
```
Comment (mirka) [inline packages/block-editor/src/components/date-format-picker/style.scss:14]: It looks the [BaseControl](https://github.com/WordPress/gutenberg/blob/71a3344e32dab008611e3196146fb736071c78ea/packages/block-editor/src/components/date-format-picker/index.js#L114) here is used merely for its implicit 8px margin-bottom properties (which is exactly the thing we're deprecating 😄). I think we can remove the BaseControl altogether and just wrap the CustomSelectControl and TextControl in a VStack.

File: packages/block-editor/src/components/url-popover/style.scss [hunk 8]
```diff
@@ -2,11 +2,13 @@
...
 
-.block-editor-url-popover__additional-controls > div[role="menu"] .components-button:not(:disabled):not([aria-disabled="true"]):not(.is-secondary) > svg {
+.block-editor-url-popover__additional-controls > div[role="menu"]
+.components-button:not(:disabled):not([aria-disabled="true"]):not(.is-secondary) > svg {
 	box-shadow: none;
 }
 
-.block-editor-url-popover__additional-controls div[role="menu"] > .components-button {
+.block-editor-url-popover__additional-controls
+div[role="menu"] > .components-button {
```
Comment (mirka) [inline packages/block-editor/src/components/url-popover/style.scss]: Just checking, are these changes intentional?

File: packages/block-library/src/video/editor.scss [hunk 9]
```diff
@@ -126,17 +109,7 @@
...
-	}
-	.components-text-control__input {
-		margin-left: 0;
-	}
-	.components-input-control__label {
-		margin-bottom: $grid-unit-05;
-	}
+.block-library-video-tracks-editor__single-track-editor-track-title,
+.block-library-video-tracks-editor__single-track-editor-source-language {
+	flex: 1;
```
Comment (mirka) [inline packages/block-library/src/video/editor.scss]: This looks like a use case for `Grid` rather than `HStack`, since we know we want two equal-width columns.

File: packages/block-library/src/video/editor.scss (hunk 5 above)
Comment (mirka) [reply packages/block-library/src/video/editor.scss]: Personally I haven't really encountered a legitimate use case for Spacer!

In this case I want to make sure we aren't mixing `margin` and `gap` in a haphazard way:

margin-top being used in addition to a flex gap

I think it's fine to space the footer row using either margin or gap, but not both 😄

File: packages/block-library/src/video/tracks-editor.js [hunk 10]
```diff
@@ -84,15 +86,20 @@ function SingleTrackEditor( { track, onChange, onClose, onRemove } ) {
 	const fileName = src.startsWith( 'blob:' ) ? '' : getFilename( src ) || '';
 	return (
 		<NavigableMenu>
-			<div className="block-library-video-tracks-editor__single-track-editor">
+			<VStack
+				className="block-library-video-tracks-editor__single-track-editor"
+				spacing="3"
```
Comment (mirka) [inline packages/block-library/src/video/tracks-editor.js]: One thing I noticed is that `spacing=3` makes it a little hard to tell that the help text is associated with the items on top. It does seem like an insufficient amount of whitespace when help text is involved. I'll leave it up to you, but this might be a good time to switch this to the new [grid spacing guidance](https://github.com/WordPress/gutenberg/issues/43423#issuecomment-1232540038), which would be `spacing=4`.

Insufficient spacing between help text for the control above and label text for the control below

File: packages/edit-site/src/components/create-template-part-modal/style.scss (hunk 3 above)
Comment (mirka) [reply packages/edit-site/src/components/create-template-part-modal/style.scss:18]: Wow this is interesting, I've never noticed it before. The standard WordPress forms.css adds a `cursor: pointer` to all `label`s:



I'm guessing the intent was to add an affordance that the labels are clickable. But unfortunately that is not always true, as is in this case (and some others, e.g. [`<progress>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/progress)). So that's probably why they explicitly overrode the `cursor: pointer` here.

cc @ciampo FYI

File: packages/edit-site/src/components/create-template-part-modal/index.js (hunk 2 above)
Comment (mirka) [reply packages/edit-site/src/components/create-template-part-modal/index.js:58]: Looks good to me 👍

**Review by mirka — COMMENTED** [review]
Good call to split the PRs!

It's nice to clean up all this CSS, but it's also nice to see the kinds of common UI patterns in the Gutenberg app. I think it will help inform our API design decisions.

File: packages/block-library/src/video/editor.scss (hunk 5 above)
Comment (brookewp) [reply packages/block-library/src/video/editor.scss]: Far enough! The reasoning here was to get closer to matching trunk. If the narrow gap is okay, I'd rather remove the margin altogether. Or add Spacer unless that's also weird to have a gap and Spacer. But then I know you'd like to avoid Spacer in general! 😅

File: packages/block-library/src/video/tracks-editor.js (hunk 10 above)
Comment (brookewp) [reply packages/block-library/src/video/tracks-editor.js]: I agree! And nice, I love consistent spacing; sooo glad to see there are guidelines! 😍

File: packages/block-library/src/video/editor.scss (hunk 9 above)
Comment (brookewp) [reply packages/block-library/src/video/editor.scss]: Awesome! I should have known there was a Grid component out there :) Thanks!

File: packages/block-editor/src/components/url-popover/style.scss (hunk 8 above)
Comment (brookewp) [reply packages/block-editor/src/components/url-popover/style.scss]: No, not at all. It must have been something weird with my formatter because I don't recognize these selectors. 😕 Sorry! 🙇‍♀️

File: packages/block-editor/src/components/date-format-picker/style.scss (hunk 7 above)
Comment (brookewp) [reply packages/block-editor/src/components/date-format-picker/style.scss:14]: Great find! I'll keep an eye out for stuff like this in the future

**Comment by brookewp** [comment]
@mirka, just to note, I also included the file `packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js` since your initial review and I've included [testing steps at the top of the testing section](https://github.com/WordPress/gutenberg/pull/47158/#HierarchicalTermSelector). 

I originally intended to include it in this PR due to some minor visual changes, but I ended up with it matching trunk. I just remembered that I needed to move it to part 1, but I merged that already. 

However, in the spirit of part 2, it would be possible to remove _most_ of the CSS (this has `align-self: flex-start;` on the 'Add New Category' button), but it would look a bit different: 
Screenshot 2023-01-17 at 4 59 32 PM
But it's an option I could include if it's preferred. :)

File: packages/block-library/src/video/editor.scss (hunk 5 above)
Comment (mirka) [reply packages/block-library/src/video/editor.scss]: Sorry I wasn't clear, we should maintain the same amount (32px) of whitespace. The point is to achieve it "logically", with a separate VStack, or a single margin-top:

The footer positioned with a margin-top, or with a separate VStack wrapper

Comment (brookewp) [reply packages/block-library/src/video/editor.scss]: Thanks for clarifying! I missed this comment earlier but my latest commit covers this now.

**Review by mirka — APPROVED** [review]
Great cleanup! 🔥 🧹

**Comment by ciampo** [comment]
Always good to see more 🔴 than 🟢 in the diff! Thank you @brookewp for working on this!
